### PR TITLE
Azure Data Share sample

### LINKED ISSUE #573: Azure DataShare automation sample
## Problem Statement

There is a lack of samples and examples on how to use [Azure Data Share](https://learn.microsoft.com/en-us/azure/data-share/overview) in an automated way. In a scenario where a company wants to share data with another company, the Data Share service makes it very easy accomplish that goal but the actual process of setting up the data share involves a lot of steps:

- creating a data share on an origin Azure Data Share account
- creating a data share on a destination Azure Data Share account
- creating a dataset to be shared on the source data share account  (ADLS as an example)
- sending an invitation to an email address or a **service principal** 
- accepting the invitation and creating a share subscription on the destination data share account - this step is manual unless a service principal is the recipient of the invitation
- mapping the incoming dataset to a destination ADLS file system

Sending an invitation to a service principal allows to fully automate the scenario on both the source and destination as we can have an Azure Function ready to accept invitations sent to a specific service principal or its managed identity.

While this is discoverable by exploring the [Invitations API](https://learn.microsoft.com/en-us/rest/api/datashare/2020-09-01/invitations/create?tabs=HTTP), it is not trivial to implement.

## Proposed Solution

A working sample of a fully automated Azure Data share process under `/single_tech_samples/datashare` that fully implements the steps outlined above and includes step-by-step documentation on how to run the code and adapt it to other scenarios.

## Additional Context

I already have a full working sample written in Python that includes simple scripts for sending/receiving invitations and an Azure Function that automates the invitation acceptance and mapping of the incoming data in a destination Data Share account. I will submit it as a PR if this is accepted.

## Next Steps

- [ ] Team consensus to proceed
- [ ] Complete Design Review Template (if applicable)
- [ ] Schedule Design Session

### PR DESCRIPTION
**Pull request by QuickNS**
# Type of PR

- Documentation changes
- Code changes

## Purpose

- Adding a sample about Azure Data Share that fully automates the share invitation process by using a service principal as the recipient of the invite.
- Includes Python scripts to originate the invite and to accept the invitation
- Sample implements an end-to-end scenario where data will be shared between two Data Share accounts and mapped to ADLS Gen2 Storage accounts.
- Includes infrastructure scripts to setup a demo environment
- Includes an Azure Function for a fully automated process that can run unattended

## Does this introduce a breaking change? If yes, details on what can break
No

## Author pre-publish checklist

- [x] Added test to prove my fix is effective or new feature works
- [x] No PII in logs
- [x] Made corresponding changes to the documentation

## Validation steps

- Sample includes step by step instructions on how to run the scenario.

## Issues Closed or Referenced

- Closes #573

### CODE REVIEW

**Review by ydaponte — CHANGES_REQUESTED** [review]
Leaving first comments for you to check. I have again an issue with the Azure Function, but will try to work it out to finish the testing and add final comments if needed.

File: single_tech_samples/datashare/Readme.md [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+- Do not include `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET` and `AZURE_TENANT_ID` in the function settings. It will default to use the managed identity for authentication and authorization.
+- Make sure the managed identity of the Azure Function has the correct permissions over Data Share account and storage account (as defined above for the service principal)
+- Modify `source.py` so that the invitation is sent to the objectId of the managed identity instead of the service principal. As a reminder, you can get this value by running the following command:
+
+  ```bash
+  az ad sp list --display-name <managed_identity_name> --query []."id" -o tsv
+  ```
+
+- Finally, ensure the managed identity is given the right permissions over the storage account and data share resources.
+- Remember that permissions can be granted by using the `bash_setup_sp_permissions.sh` or `ps_setup_sp_permissions.sh` script in the `infra` folder.
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:406]: Added a section at the end with cleanup instructions

File: single_tech_samples/datashare/Readme.md [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    - [Azure Function requirements](#azure-function-requirements)
+    - [F5 experience](#f5-experience)
+    - [Azure function authentication](#azure-function-authentication)
+
+## Description
+
+[Azure Data Share](https://azure.microsoft.com/en-us/services/data-share/) setup requires a number of steps to establish the connection between the source data and the destination. One of those steps is sending an invitation from a source data share account and accepting the invitation in a destination data share account.
+
+Through the portal UI, invitations can only be sent to email addresses and that requires the email recipient to perform some manual steps to accept the invitation and map the incoming data to the destination. However, the Azure Data Share SDK allows invitations to be sent to *service principals* as well, which opens up the opportunity to fully automate the process, even between different subscriptions and tenants.
+
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:33]: Added a "Using the sample" section

File: single_tech_samples/datashare/Readme.md [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+- Enables the scheduling trigger
+
+### Azure Function
+
+An Azure Function with a timer trigger is included, so the acceptance of invitations can be fully automated instead of on-demand.
+This code can be found on the `azure_function` folder but it's recommended that the scripts are used for initial debugging and testing.
+
+### Dev Container
+
+A dev container is included to allow for an easier setup of your local dev environment. It includes all the required dependencies to execute the code and is the recommended way of testing this sample. For more information, please refer to <https://code.visualstudio.com/docs/devcontainers/containers> for more information.
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:59]: Nice catch. Fixed!

File: single_tech_samples/datashare/Readme.md [hunk 4]
```diff
@@ -0,0 +1,406 @@
+# Data Share Automation
+
+- [Data Share Automation](#data-share-automation)
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:3]: Moved the table of contents to a new "Contents" section that only includes the remaining topics in the doc

File: single_tech_samples/datashare/Readme.md [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+## Role Assignments
+
+- The **source** data share MSI must have the **Storage Blob Data Reader** role in the source storage account
+
+    Example:
+    ![add-data-share-permissions](./media/add-data-share-permissions.png)
+
+- The **destination** data share MSI must have the **Storage Blob Data Contributor** role in the destination storage account.
+
+If you used the *infra* scripts in this repo to create the resources, the role assignements have been setup automatically.
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:145]: One of the challenges of writing this doc is trying to ensure instructions work in both cases (using the infra scripts or trying with pre-existing assets, or manually created assets). After a bit of back and forth, I decided this iterative approach where settings and roles are mentioned when they are needed in the tutorial steps works better than trying to list them upfront. So I'll just move the line and rephrase it a bit.

File: single_tech_samples/datashare/Readme.md [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+az ad sp list --display-name <insert_sp_name> --query []."id" -o tsv
+```
+
+These values will be referenced in the scripts through environment variables:
+
+- AZURE_CLIENT_ID: appId
+- AZURE_CLIENT_SECRET: password
+- AZURE_TENANT_ID: tenant
+
+The objectId will be used in the `source.py` script as the target of the invitation. You will need to add the object_id
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:134]: yes, this phrase was clearly left incomplete, so I just rephrased this whole section.

File: single_tech_samples/datashare/Readme.md [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+
+If you want to automatically assign these permissions, you can use the `bash_setup_sp_permissions.sh` or `ps_setup_sp_permissions.sh` script in the `infra` folder:
+
+```bash
+# login
+az login
+# make sure the right subscription is selected
+az account set -s <SUBSCRIPTION_ID>
+# add permissions for the service principal objectId
+./infra/bash_setup_sp_permissions.sh <suffix> <objectId>
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md:305]: it works both ways apparently.

File: single_tech_samples/datashare/Readme.md [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+Execute the following commands:
+
+```bash
+cd python
+pip install -r requirements.txt
+
+python dest.py
+```
+
+After the invitation is accepted, the script can't be run again. If you need to re-run it, please create a new invitation using the `dest.py` python script.
```
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md]: yes, fixed

File: single_tech_samples/datashare/Readme.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuickNS) [reply single_tech_samples/datashare/Readme.md]: Following your suggestion, I restructured the document a bit. Please review.

**Review by ydaponte — APPROVED** [review]
Readability improved a lot with the changes. Thanks for doing them. I'm approving the PR now.

**Pull request merged**